### PR TITLE
C procedural gen

### PR DESCRIPTION
**Pull request by ZynoXelek**
C-porting the code for enhanced performances and control over data storage and usage.

### CODE REVIEW

**Comment by ebrossier-secher** [comment]
Everything is valgrind-friendly until now!

**Comment by ebrossier-secher** [comment]
Need to modify map generation (change has to be done in chunk structure) to stop affecting structure list variable to a made copy of the list to avoid freeing issues.
Except for this, everything works fine and is valgrind-friendly until now.

**Comment by ebrossier-secher** [comment]
Everything is valgrind-friendly.
Better signatures for functions.
Lists in arguments are now copied in dynamically allocated arrays, and there are no more issues with different chunks pointing to the same array of factors_layers.

No allocation are required in the testing scripts nor the future main scripts.

Still needed :
- [ ] Functions to print completeMap structure ?
- [x] Functions to save structure in files
- [x] Python (or C script) to plot the saved data

And obviously :
- [ ] COMMENTS !!!

I remove the draft tag. The code is ready for a pre-review of the already implemented functions.

File: C/map.c [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    clock_t start_time = clock();
+
+    if (chunks != NULL && map_width > 0 && map_height > 0)
+    {
+        map* new_map = calloc(1, sizeof(map));
+
+        new_map->map_width = map_width;
+        new_map->map_height = map_height;
+
+        // copy chunks list to ensure dynamic allocation
```
Comment (BlueNZ) [inline C/map.c:73]: I am not sure I follow why you need to do so, and why a simple list copy rather than a 'deep copy'

Comment (BlueNZ) [reply C/map.c:73]: Nevermind I understood

**Review by BlueNZ — APPROVED** [review]
I only reviewed the file @ZynoXelek didn't present me last time (for my time is precious), and as far as I can see all seems fair and fine. 
I may simply regret the lack of comments to explain how to use each function and how they work internally. But as I, myself, tend to forget those things, it is mostly inconsequential.

**Comment by ebrossier-secher** [comment]
> I only reviewed the file @ZynoXelek didn't present me last time (for my time is precious), and as far as I can see all seems fair and fine. I may simply regret the lack of comments to explain how to use each function and how they work internally. But as I, myself, tend to forget those things, it is mostly inconsequential.

Yes indeed, comments are clearly MISSING and REQUIRED! I shall add them soon.

**Comment by ebrossier-secher** [comment]
I made a test to compare python generation and C generation. You can look it yourself @Laudut and @BlueNZ but, on my computer and without printing loading bars nor plotting the final graph (for the python version) and without printing nor generating the files (for the C version) :

For a generation of a map with 3 layers and size 5 x 7 :
- The python code takes 9.094642877578735 seconds
- The C code takes 0.132694 seconds

Which makes the C code around 68,38 times faster.


For the generation itself, you can load the files generated by the test_mapGenerator.c file with the main.py by setting `LOAD = True`.
We do not obtain the exact same map as the Python generated map, certainly due to the way seeds work on both languages.
Moreover, there may be some mistakes in the C generation because it seems to have way less water with the same water level, and the general plotting seams darker.
It needs to be tested.

Finally, it still lacks comments, but I wanted to solve this first.

**Comment by ebrossier-secher** [comment]
Would still need more comments but ready for a technical merge.

**Pull request merged**